source/client/service_impl.cc

```diff
@@ -67,20 +67,22 @@ ::grpc::Status ServiceImpl::ExecutionStream(
                                ::nighthawk::client::ExecutionRequest>* stream) {
   nighthawk::client::ExecutionRequest request;
   stream_ = stream;
-  busy_ = false;
   while (stream->Read(&request)) {
     ENVOY_LOG(debug, "Read ExecutionRequest data: {}", request.DebugString());
     if (request.has_start_request()) {
       // It is possible to receive a back-to-back request here, while the future that is associated
       // to our previous response is still active. We check the running_ flag to see if the previous
       // future has progressed in a state where we can do another one. This avoids the odd flake in
       // ServiceTest.BackToBackExecution.
-      if (busy_) {
+      if (future_.valid() &&
+          future_.wait_for(std::chrono::seconds(0)) != std::future_status::ready && busy_) {
```

Comment:
Looks good, thanks. One last question; why do we still need `busy_`? Isn't knowing that the future isn't ready sufficient to know that it's still in-flight?

Reply:
I think the thing is that the future may have send the reply and still not be ready when the next request comes in. So we may even receive the next request before the future has torn down.
To illustrate: the print I added in the following diff will be seen in the out when reproducing the failure in parallell/repeated tests (note that the busy flag is no longer checked):

```diff
diff --git a/source/client/service_impl.cc b/source/client/service_impl.cc
index fca8290..853cb77 100644
--- a/source/client/service_impl.cc
+++ b/source/client/service_impl.cc
@@ -75,7 +75,8 @@ void ServiceImpl::writeResponse(const nighthawk::client::ExecutionResponse& resp
       // future has progressed in a state where we can do another one. This avoids the odd flake in
       // ServiceTest.BackToBackExecution.
       if (future_.valid() &&
-          future_.wait_for(std::chrono::seconds(0)) != std::future_status::ready && busy_) {
+          future_.wait_for(std::chrono::seconds(0)) != std::future_status::ready) {
+        std::cerr << "expected to not be busy" << std::endl;
         return finishGrpcStream(false, "Only a single benchmark session is allowed at a time.");
       } else {
         busy_ = true;
```

Reply:
Yes, that seems entirely possible. I think `busy_` is OK to unblock the import, but it's also super icky IMHO, since we're doing manual concurrency control via atomics. Ideally we would have something more like a giant select on the message loop, rather than a gRPC `Read`. Combining futures and gRPC APIs is kind of messy.Did we explore maybe using something like https://en.cppreference.com/w/cpp/thread/mutex/try_lock and a mutex?